### PR TITLE
App platform: Implement perm check with direct db access

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Implementing permission check using direct DB access, so we can check RBAC permissions without requiring the instance to be running.

Some implementation details:
* uses SQL templates;
* fetches permissions associated with the given action and searches through them;
* takes wildcards into account.

Needs to be covered in the future:
* resource listing;
* permission inheritance (from the parent folder and parent's parents);
* permission caching;
* perm checks for any other identity types than user and service account (eg, anonymous auth)

**Who is this feature for?**

Internal

**Which issue(s) does this PR fix?**:

Related to https://github.com/grafana/identity-access-team/issues/1022

**Special notes for your reviewer:**

See description above for what is not covered in this PR.

I've tested the current changes locally and will also add unit tests for them.

### CODE REVIEW

**Review by gamab — APPROVED** [review]
Great job. LGTM.

We need to check for the caller namespace.

Among the things that we'll need to clarify:
* OSS has roles in RAM, should we handle that case?
* Are checks ever going to be done in the global organization?

File: pkg/services/authz/rbac/store/permission_query.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+SELECT p.action, p.kind, p.attribute, p.identifier, p.scope from {{ .Ident .PermissionTable }} as p
+where p.action = {{ .Arg .Query.Action }} AND p.role_id in (
+  SELECT role_id from {{ .Ident .UserRoleTable }} as ur where ur.user_id = {{ .Arg .Query.UserID }} AND (ur.org_id = {{ .Arg .Query.OrgID }} OR ur.org_id = 0)
+  {{ if .Query.TeamIDs }}
+    UNION
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/authz/rbac/store/permission_query.sql:5]: Yes, as roles that are assigned to user and users teams directly can overlap (so I'm using UNION), but will never contain basic roles (so I'm using UNION ALL for those). But I don't think it makes any significant difference performance wise, so I'm also happy to switch to using only UNION if you prefer that.

File: pkg/services/authz/rbac/store/store.go [hunk 2]
```diff
@@ -13,3 +20,73 @@ func NewStore(sql legacysql.LegacyDatabaseProvider) *Store {
...
+	defer func() {
+		if rows != nil {
+			_ = rows.Close()
+		}
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	if !rows.Next() {
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/authz/rbac/store/store.go:82]: Ah, good catch. I'll test this scenario.

[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/authz/rbac/store/store.go:116]: I've changed the code to account for this use case - I'm assuming that if a user is not a member of the specified org, they have `None` permissions in this org.

File: pkg/services/authz/rbac/service.go [hunk 3]
```diff
@@ -14,24 +25,178 @@ type Service struct {
...
+	if err != nil {
+		ctxLogger.Error("could not parse namespace", "namespace", namespace, "error", err)
+		return nil, err
+	}
+
+	if req.GetSubject() == "" {
+		return nil, status.Error(codes.InvalidArgument, "subject is required")
+	}
+	user := req.GetSubject()
+	identityType, userUID, err := claims.ParseTypeID(user)
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/authz/rbac/service.go:92]: I've rewritten the identifier resolver logic (I realised that the other implementation also doesn't allow resolving service account UIDs, so it was not suitable for our use case anyway).

**Pull request merged**